ulib/environ.c

```diff
@@ -0,0 +1,250 @@
+#include "fcntl.h"
+#include "file_def.h"
+#include "stdio.h"
+#include "stdlib.h"
+#include "string.h"
+#include "unistd.h"
+#include "errno.h"
+#include "../kernel/param.h"
+
+#define PATH_ENV_VAR "PATH"
+#define ENV_VAR_LEN   (128)
+#define ENV_VAR_DELIM   '='
+
+char **environ={0};
+extern int _path_update(const char *);
+
+typedef struct __evar_list{
+  struct __evar_list *prev;
+  struct __evar_list *next;
+}_evar_list;
+
+typedef struct __evar_head{
+  struct __evar_list *prev;
+  struct __evar_list *next;
+}_evar_head;
+
+typedef struct _evar_type{
+  _evar_list link;
+  char       *var;
+  char     *namep;
+  char      *valp;
+}evar_type;
+
+static evar_type evar_array[MAXENV];
+static _evar_head evars={.prev = (struct __evar_list *)&evars, .next = (struct __evar_list *)&evars};
+
+static int
+env_add_from_environ(const char *var, int overwrite){
+  int i;
+  evar_type *p;
+  int  len;
+
+  for(i=0; MAXENV>i; ++i) {
+
+    p = &evar_array[i];
+		
+    if ( p->var != 0 ) {
```

Comment:
Can you use `NULL` instead of `0`?

Reply:
Sorry, I have a question.
Can we use NULL in kernel codes?
Apparently, xv6 does not use NULL in kernel codes because it is designed to provide a "flavor" of Unix version 6. So it does not use NULL. I would also like to use NULL also and it is preferable to improve code maintainability.

Reply:
Files under ulib directory are user library,
so I think its reasonable to use `NULL`.